pkg/util/provider/machinecontroller/machine_util.go

```diff
@@ -389,52 +392,102 @@ func SyncMachineTaints(
+	clone.Status.LastOperation = lastOperation
+	clone.Status.CurrentStatus = currentStatus
+	clone.Status.LastKnownState = lastKnownState
+
+	if isMachineStatusSimilar(clone.Status, machine.Status) {
+		klog.V(3).Infof("Not updating the status of the machine object %q, as the content is similar", clone.Name)
+		return nil
 	}
 
-	_, err = c.controlMachineClient.Machines(clone.Namespace).UpdateStatus(clone)
+	_, err := c.controlMachineClient.Machines(clone.Namespace).UpdateStatus(clone)
 	if err != nil {
 		// Keep retrying until update goes through
-		klog.Errorf("Machine/status UPDATE failed for machine %q. Retrying, error: %s", machine.Name, err)
+		klog.Warningf("Machine/status UPDATE failed for machine %q. Retrying, error: %s", machine.Name, err)
 	} else {
-		klog.V(2).Infof("Machine/status UPDATE for %q during CREATE error", machine.Name)
+		klog.V(2).Infof("Machine/status UPDATE for %q", machine.Name)
 	}
 
-	return retryRequired, err
+	return err
+}
+
+// isMachineStatusSimilar checks if the status of 2 machines is similar or not.
+func isMachineStatusSimilar(s1, s2 v1alpha1.MachineStatus) bool {
+	s1Copy, s2Copy := s1.DeepCopy(), s2.DeepCopy()
+	tolerateTimeDiff := 30 * time.Minute
+
+	// Since lastOperation hasn't been updated in the last 30minutes, force update this.
+	if (s1.LastOperation.LastUpdateTime.Time.Before(time.Now().Add(tolerateTimeDiff * -1))) || (s2.LastOperation.LastUpdateTime.Time.Before(time.Now().Add(tolerateTimeDiff * -1))) {
+		return false
+	}
+
+	if utilstrings.StringSimilarityRatio(s1Copy.LastOperation.Description, s2Copy.LastOperation.Description) > 0.75 {
+		// If strings are similar, ignore comparison
+		// This occurs when cloud provider errors repeats with different request IDs
+		s1Copy.LastOperation.Description, s2Copy.LastOperation.Description = "", ""
+	}
+
+	// Avoiding timestamp comparison
+	s1Copy.LastOperation.LastUpdateTime, s2Copy.LastOperation.LastUpdateTime = metav1.Time{}, metav1.Time{}
+	s1Copy.CurrentStatus.LastUpdateTime, s2Copy.CurrentStatus.LastUpdateTime = metav1.Time{}, metav1.Time{}
+
+	return apiequality.Semantic.DeepEqual(s1Copy.LastOperation, s2Copy.LastOperation) && apiequality.Semantic.DeepEqual(s1Copy.CurrentStatus, s2Copy.CurrentStatus)
+}
+
+// getCreateFailurePhase gets the effective creation timeout
+func (c *controller) getCreateFailurePhase(machine *v1alpha1.Machine) v1alpha1.MachinePhase {
+	timeOutDuration := c.getEffectiveCreationTimeout(machine).Duration
+	// Timeout value obtained by subtracting last operation with expected time out period
+	timeOut := metav1.Now().Add(-timeOutDuration).Sub(machine.CreationTimestamp.Time)
+
+	if timeOut > 0 {
+		// Machine creation timeout occured while joining of machine
+		// Machine set controller would replace this machine with a new one as phase is failed.
+		klog.V(2).Infof("Machine %q couldn't join in creation timeout of %s. Changing phase to failed.", machine.Name, timeOutDuration)
```

Comment:
This would mean the creation-timeout is now respected even with CrashLoopBackOff. Thanks for taking care of it.

Reply:
Yes. That is right. This would help providers where they want to retry from scratch after some time.